### PR TITLE
[BEAM-6929] Prevent NullPointerException in Flink's CombiningState

### PR DESCRIPTION
**Pull request by mxm**
When the accumulator is retrieved and it has not been initialized yet, the
Flink's StateInternals return null. This can lead to a NullPointerException for
Session windows with allowed lateness, where the state is cleared and the
accumulator is retrieved afterwards.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by mxm** [comment]
CC @tweise @angoenka

**Comment by mxm** [comment]
Appears this is a bug in the Spark Runner as well..

**Comment by kennknowles** [comment]
Reading the code, I wondered if it might be pervasive in most/all runners. Is it a release blocker though? This bug has existed presumably for many releases.

**Comment by mxm** [comment]
>Is it a release blocker though? This bug has existed presumably for many releases.

I'd say yes because it is reported as a blocker for session windows by a user on the mailing list. I've verified that it prevents merging session windows when late data arrives.

The bug in Spark's state internals was easy to fix. It assumed that accumulators are always modified in place, which does not have to be the case (though it usually is for performance reasons).

File: runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/SparkStateInternals.java [hunk 1]
```diff
@@ -297,8 +297,7 @@ public OutputT read() {
 
     @Override
     public void add(InputT input) {
-      AccumT accum = getAccum();
-      combineFn.addInput(accum, input);
+      AccumT accum = combineFn.addInput(getAccum(), input);
```
Comment (mxm) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/SparkStateInternals.java]: CC @iemejia FYI

**Comment by mxm** [comment]
Run Java PreCommit

**Comment by iemejia** [comment]
Two extra questions @mxm:
1. Does this mean that if they are not calculated in place we can have a future consistency issue for spark?
2. Did you check in other runners too?
This is worth the cherry-pick indeed.

**Comment by mxm** [comment]
Thanks for merging!

>Does this mean that if they are not calculated in place we can have a future consistency issue for spark?

Yes, combiners which do not work in-place but on a new accumulator produced wrong results for combining state. That said, I think almost all combiners work in-place.

>Did you check in other runners too?

I adjusted the `StateInternalsTest` which Flink, Spark, Apex, and the InMemoryStateInternals use.

**Pull request merged**